pkg/dumper/postgres/dumper.go

```diff
@@ -0,0 +1,190 @@
+package postgres
+
+import (
+	"database/sql"
+	"sync"
+
+	"github.com/hellofresh/klepto/pkg/config"
+	"github.com/hellofresh/klepto/pkg/database"
+	"github.com/hellofresh/klepto/pkg/dumper"
+	"github.com/hellofresh/klepto/pkg/reader"
+	"github.com/lib/pq"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// pgDumper dumps a database into a postgres db
+type pgDumper struct {
+	conn   *sql.DB
+	reader reader.Reader
+}
+
+func NewDumper(conn *sql.DB, rdr reader.Reader) dumper.Dumper {
+	return &pgDumper{
+		conn:   conn,
+		reader: rdr,
+	}
+}
+
+func (p *pgDumper) Dump(done chan<- struct{}, configTables config.Tables) error {
+	if err := p.dumpStructure(); err != nil {
+		return err
+	}
+
+	return p.dumpTables(done, configTables)
+}
+
+func (p *pgDumper) Close() error {
+	return p.conn.Close()
+}
+
+func (p *pgDumper) dumpStructure() error {
+	log.Debug("Dumping structure...")
+	structureSQL, err := p.reader.GetStructure()
+	if err != nil {
+		return err
+	}
+
+	_, err = p.conn.Exec(structureSQL)
+	log.Debug("Structure dumped")
+
+	return err
+}
+
+func (p *pgDumper) dumpTables(done chan<- struct{}, configTables config.Tables) error {
+	// Get the tables
+	tables, err := p.reader.GetTables()
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(tables))
+	for _, tbl := range tables {
+		var opts reader.ReadTableOpt
+
+		table, err := configTables.FindByName(tbl)
+		if err != nil {
+			log.WithError(err).WithField("table", tbl).Debug("no configuration found for table")
+		}
+
+		if table != nil {
+			opts = reader.ReadTableOpt{
+				Limit:         table.Filter.Limit,
+				Relationships: p.relationshipConfigToOptions(table.Relationships),
+			}
+		}
+
+		// Create read/write chanel
+		rowChan := make(chan database.Row)
+
+		go func(tableName string, rowChan <-chan database.Row) {
+			if err := p.dumpTable(tableName, rowChan); err != nil {
+				log.WithError(err).WithField("table", tableName).Error("Failed to dump table")
+			}
+
+			wg.Done()
+		}(tbl, rowChan)
+
+		p.reader.ReadTable(tbl, rowChan, opts)
+	}
+
+	go func() {
+		// Wait for all table to be dumped
+		wg.Wait()
```

Comment:
This will not block as it's inside a goroutine. We want this to block since you are waiting for the other goroutines to be done?

Reply:
This line blocks `done <- struct{}{}` which is correct because we are done when all tables are dumped :smile: or my understanding must be going into the :tropical_drink: mode :smile: